src/core_plugins/kibana/public/management/sections/settings/advanced_import_controls.js

```diff
@@ -0,0 +1,93 @@
+import { saveAs } from '@spalger/filesaver';
+import { AdvancedImportState } from 'plugins/kibana/management/sections/settings/advanced_import_state.js';
+import advancedImportControlsTemplate from 'plugins/kibana/management/sections/settings/advanced_import_controls.html';
+
+uiModules.get('apps/management')
+  .directive('advancedImportControls', function (config, Notifier, Private) {
+    const notify = new Notifier({ location: 'Advanced Settings' });
+    const advancedImportState = Private(AdvancedImportState);
+
+    return {
+      restrict: 'E',
+      template: advancedImportControlsTemplate,
+      controllerAs: 'advancedImportControls',
+      bindToController: true,
+      controller: class AdvancedImportController {
+        constructor() {
+          this.importState = advancedImportState;
+
+          this.export = () => {
+            const all = config.getAll();
+            // Construct an object with keys that have user-defined values and values
+            // set to those user-defined values
+            const userValues = Object.keys(all).reduce((values, configKey) => {
+              const configObj = all[configKey];
+
+              // This logic is a bit messy. We only want user-defined values that are configurable
+              // so that immediately excludes readonly values
+              if (configObj.readonly) {
+                return values;
+              }
+
+              // Now, we need to make sure they have a user-defined value
+              if (!configObj.hasOwnProperty('userValue')) {
+                return values;
+              }
+
+              // It is also possible that the user-defined value matches
+              // the default value
+              if (configObj.userValue === configObj.value) {
+                return values;
+              }
+
+              values[configKey] = configObj.userValue;
+              return values;
+            }, {});
+            const blob = new Blob([JSON.stringify(userValues)], { type: 'application/json' });
+            saveAs(blob, 'settings-export.json');
+          };
+
+          this.import = (fileContents) => {
+            let settings;
+            try {
+              settings = JSON.parse(fileContents);
+            } catch (e) {
+              notify.error('The file could not be processed.');
+              return;
+            }
+
+            advancedImportState.clearSettings();
+
+            const allConfig = config.getAll();
+            Object.keys(settings).forEach(settingName => {
+              const setting = settings[settingName];
+              const isValid = allConfig.hasOwnProperty(settingName);
+              if (isValid) {
+                // Do not bother showing if they imported values match the current values
+                if (setting === config.get(settingName)) {
+                  return;
+                } else if (JSON.stringify(setting) === JSON.stringify(config.get(settingName))) {
```

Comment:
There's a small issue with this comparison, due to whitespace. A simple fix for this would be to simply remove whitespace (and probably newlines) before doing the comparison.

![screenshot 2017-05-31 13 36 19](https://cloud.githubusercontent.com/assets/404731/26652937/680b65e2-4606-11e7-838a-befdfa4818ef.png)